webapp/app.py

```diff
@@ -385,3 +388,68 @@ def soc_models():
         page=page,
         pages=get_pagination_page_array(page, num_pages),
     )
+
+
+@app.route("/components")
+def components():
+    query = flask.request.args.get("query") or ""
+    page = int(flask.request.args.get("page") or "1")
+    vendors = flask.request.args.getlist("vendor")
+
+    components_response = api.componentsummaries(
+        vendor_name__in=",".join(vendors) if vendors else None,
+        # We should use query instead of make__regex as soon as it's ready
+        # query=query,
+        model__regex=query,
+        offset=(int(page) - 1) * 20,
+    )
+    components = components_response["objects"]
+    total = components_response["meta"]["total_count"]
+
+    num_pages = math.ceil(total / 20)
+
+    all_vendors = ["AMD", "Lenovo", "nVidia"]
+
+    params = flask.request.args.copy()
+    params.pop("page", None)
+    query_items = []
+    for key, valuelist in params.lists():
+        for value in valuelist:
+            query_items.append(f"{key}={value}")
+
+    return flask.render_template(
+        "components/index.html",
+        components=components,
+        query=query,
+        vendors=vendors,
+        all_vendors=sorted(all_vendors),
+        total=total,
+        query_string="&".join(query_items),
+        page=page,
+        pages=get_pagination_page_array(page, num_pages),
+    )
+
+
+@app.route("/components/<id>")
+def component_details(id):
+    component = api.componentsummary(id)
+
+    all_machines = api.certifiedmodels(
+        canonical_id__in=component["machine_canonical_ids"]
+    )["objects"]
+
+    machines_by_id = {}
```

Comment:
Is this because the API returns duplicates?

Reply:
yep